### PR TITLE
Feat: Add custom media queries

### LINKED ISSUE #2: Custom Media Queries
as of now, this linter doesn't allow custom @media queries that Bootstrap doesn't support (https://getbootstrap.com/docs/5.0/layout/breakpoints/) to have CSS classes that would be disallowed.

### PR DESCRIPTION
**Pull request by Spayco**
I added a check to custom Bootstrap media queries, so the lint would only work if the properties are in the root or are inside those media queries,
all queries can be found on `Index.js` in the `bootstrapQueries` variable.

I also added two tests for them to test positive and negative scenarios.

Fixes #2

### CODE REVIEW

File: disallowedProperties.js [hunk 1]
```diff
@@ -76,6 +76,7 @@ const properties = {
 	},
 	'flex-wrap': {
 		values: {
+			auto: 'flex-auto',
```
Comment (bolah2009) [inline disallowedProperties.js:79]: Can we add this as a separate PR? This seems to be adding more bootstrap class which is a good case for a new PR.

Also, I don't think there's a bootstrap class called `flex-auto`, see https://github.com/twbs/bootstrap/blob/ed0fc832166b0911f1534a2dab7ae38f01d5a119/dist/css/bootstrap.css#L6496-L6506

File: .eslintignore [hunk 2]
```diff
@@ -1 +1,2 @@
+
```
Comment (bolah2009) [inline .eslintignore:1]: Can we remove this empty line?

File: index.js [hunk 3]
```diff
@@ -27,22 +27,37 @@ const rule = stylelint.createPlugin(ruleName, (primaryOption) => {
...
+				'max-width: 1399.98px',
+			];
+
+			if (
+				type === 'root' ||
+				(type === 'atrule' && bootstrapQueries.some((element) => params.includes(element)))
+			) {
+				if (selector.includes(':')) {
+					return;
+				}
```
Comment (bolah2009) [inline index.js:46]: Rather than using nested ifs, let's create a function that checks if a declaration has a parent with this criteria if a declaration does not satisfy it, we `return;` like how the existing code does for `selector.includes(':')`

**Review by bolah2009 — COMMENTED** [review]
Awesome job with the PR! Thanks for opening this.

There are some cases we may need to consider:
- when multiple media query is used, e.g. `@media screen and (max-width: 576px)` or `@media screen and (max-width: 576px) and (min-width: 200px)`
- what happens when a close media query is used, e.g `(max-width: 576px)`

File: __tests__/index.js [hunk 4]
```diff
@@ -53,6 +53,13 @@ testRule({
 			code: 'div::before { display: flex; }',
 			description: 'When property is disallowed but uses psuedo element selector',
 		},
+		{
+			code: "@media ('max-width: 800px') { div { display: flex; } }",
+			message: messages.rejected('display', 'flex', 'd-flex'),
+			line: 1,
+			column: 40,
+			description: 'When property is disallowed but media query is allowed',
+		},
```
[the remark this answers is not in the document]
Comment (Spayco) [reply __tests__/index.js:62]: I removed the rejected message, I mistakenly left it when I copied the test case

File: __tests__/index.js [hunk 5]
```diff
@@ -42,7 +42,7 @@ testRule({
 			description: 'When property is allowed',
 		},
 		{
-			code: 'a { top: 0; color: pink; }',
+			code: 'a { top: 2.5; color: pink; }',
```
[the remark this answers is not in the document]
Comment (Spayco) [reply __tests__/index.js:45]: I changed this value because it makes the test fail since bootstrap has `top: 0;`

**Review by bolah2009 — COMMENTED** [review]
I leave some comments.

File: index.js [hunk 6]
```diff
@@ -10,6 +10,40 @@ const messages = stylelint.utils.ruleMessages(ruleName, {
...
+		'max-width: 1199.98px',
+		'max-width: 1399.98px',
+	];
+
+	if (type === 'root' || name !== 'media') {
+		return false;
+	}
+
+	// the regex outputs an array of all numbers in the sting.
+	const numbers = params.match(/(-\d+|\d+)(,\d+)*(\.\d+)*/g);
```
[the remark this answers is not in the document]
Comment (Spayco) [reply index.js:27]: it would, I thought it would be better to cover those as well, instead of only `max-width`

[the remark this answers is not in the document]
Comment (Spayco) [reply index.js:27]: do you prefer to make it only work on `max-width` @bolah2009?

File: index.js [hunk 7]
```diff
@@ -10,6 +10,40 @@ const messages = stylelint.utils.ruleMessages(ruleName, {
...
+
+	if (numbers.length === 1) {
```
[the remark this answers is not in the document]
Comment (Spayco) [reply index.js:29]: since on Bootstrap they don't cover breakpoints with two numbers, I figured most of the time when students would have 2 numbers, it would be because it doesn't work on Bootstrap.

File: index.js [hunk 8]
```diff
@@ -27,6 +61,10 @@ const rule = stylelint.createPlugin(ruleName, (primaryOption) => {
 
 			const { selector } = parent;
 
+			if (queryChecker(parent.parent)) {
```
[the remark this answers is not in the document]
Comment (Spayco) [reply index.js:64]: I tried doing { parent }, but since it was declared already, it gives errors

File: index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (bolah2009) [reply index.js]: What do you think?

[the remark this answers is not in the document]
Comment (Spayco) [reply index.js]: @bolah2009 I'm neutral to the idea, I guess we could make it configured so it either only uses max-width or all of them? (and maybe even custom ones)

[the remark this answers is not in the document]
Comment (bolah2009) [reply index.js]: Okay, let's do it! Make it specific to `max-width` only and also document it.

Others should be a warning, not an error.

File: __tests__/index.js [hunk 9]
```diff
@@ -53,6 +53,43 @@ testRule({
...
 			description: 'When property is disallowed but uses psuedo element selector',
 		},
+		{
+			code: "@media ('max-width: 850px') { div { display: flex; } }",
+			description: 'When property is disallowed but media query is allowed',
+		},
+		{
+			code:
+				"@media screen and ('max-width: 800px') and ('min-width: 200px'){ div { display: flex; } }",
+			description: 'When property is disallowed but media query is allowed',
```
Comment (bolah2009) [inline __tests__/index.js:63]: Can we make this test description more descriptive and different?

File: __tests__/index.js [hunk 10]
```diff
@@ -53,6 +53,43 @@ testRule({
...
+			code: "@media ('max-height: 670px') { div { display: flex; } }",
+			description: 'When property and its media query is allowed',
+		},
+		{
+			code: "@media ('min-height: 670px') { div { display: flex; } }",
+			description: 'When property and its media query is allowed',
+		},
+		{
+			code: "@media ('height: 670px') { div { display: flex; } }",
+			description: 'When property and its media query is allowed',
```
Comment (bolah2009) [inline __tests__/index.js:87]: Same with these too.

**Review by bolah2009 — DISMISSED** [review]
LGTM 🚀 

Please make the changes before merging.

File: README.md [hunk 11]
```diff
@@ -49,6 +49,8 @@ This rule ignores:
 - CSS pseudo-elements i.e. those prefixed with a single colon
 - CSS pseudo-classes i.e. those prefixed with two colon
 - selectors that use variable interpolation e.g. `::#{$variable} {}`
+- Media Queries that are 50px more or less than the breakpoints used on Bootstrap
+- Media Queries that are note `max-width`
```
[the remark this answers is not in the document]
Comment (Spayco) [reply README.md:53]: oh yes, fixed!

**Comment by Spayco** [comment]
@bolah2009 I still need you to approve it before I can merge, I enabled auto-merge so once it's approved it will be merged automatically

**Review by bolah2009 — APPROVED** [review]
